### PR TITLE
Switched to k8s 1.18.0 as attempt to fix issue #25

### PR DESCRIPTION
**Pull request by petermicuch**
This is my very first attempt to work with GO at all, so it definitely needs someone more experienced to check. I only installed GO today to fix this.

As right now I have no k8s cluster running version 1.20 I can only test it tomorrow. My main goal was to update dependencies and interfaces so that this runs against k8s 1.18. With that the problem described in issue #25 shall not be present. Actually already version 1.16 should do according to [this post](https://github.com/kubernetes/kubernetes/issues/94660#issuecomment-689887045).

### CODE REVIEW

**Comment by petermicuch** [comment]
I just tested this now and it works. Volume gets created, after PVC deletion volume is removed and respective folder is marked as archived. I have to check with my company if it is fine to sign CLA.

File: cmd/nfs-subdir-external-provisioner/provisioner.go [hunk 1]
```diff
@@ -134,10 +134,10 @@ func (p *nfsProvisioner) Provision(options controller.ProvisionOptions) (*v1.Per
 			},
 		},
 	}
-	return pv, nil
+	return pv, controller.ProvisioningFinished, nil
 }
 
-func (p *nfsProvisioner) Delete(volume *v1.PersistentVolume) error {
+func (p *nfsProvisioner) Delete(ctx context.Context,volume *v1.PersistentVolume) error {
```
Comment (jackielii) [inline cmd/nfs-subdir-external-provisioner/provisioner.go:140]: runt `go fmt ./...` at the project root

File: cmd/nfs-subdir-external-provisioner/provisioner.go [hunk 2]
```diff
@@ -185,15 +185,15 @@ func (p *nfsProvisioner) Delete(volume *v1.PersistentVolume) error {
 }
 
 // getClassForVolume returns StorageClass
-func (p *nfsProvisioner) getClassForVolume(pv *v1.PersistentVolume) (*storage.StorageClass, error) {
+func (p *nfsProvisioner) getClassForVolume(context context.Context, pv *v1.PersistentVolume) (*storage.StorageClass, error) {
```
Comment (jackielii) [inline cmd/nfs-subdir-external-provisioner/provisioner.go:188]: change param name `context` to `ctx`

**Comment by petermicuch** [comment]
I have to do small internal training in my company to be able to sign CLA, so it will take a week or so before I can sign it (I hope not more). If anyone is willing to create another PR bringing in changes or even to bump directly to k8s 1.20.0 I am fine with that. I took versions of client-go as I have seen them in [go.mod](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/blob/920c708218be0fadbc50fea4d25b42c279fefe21/go.mod#L15) originally.

**Comment by rkevin-arch** [comment]
Thank you so much for the PR! It fixed our cluster for now. For anyone still using the (unfortunately deprecated but no replacement available) `stable/nfs-client-provisioner` helm chart, you can add the following to the helm values:
```yaml
image:
  repository: rkevin/nfs-subdir-external-provisioner
  tag: fix-k8s-1.20
```
This image is directly built on this PR. This is a temporary fix until this PR is merged upstream.

**Comment by petermicuch** [comment]
Almost there with CLA signing. My company representative will sign corporation CLA hopefully today.

**Comment by petermicuch** [comment]
CLA signed. Please check.

**Comment by petermicuch** [comment]
@kmova can this one be approved before #29 or will this come after? And do you know if @ashishranjan738 is around for reviews or could you help review this one?

**Comment by kmova** [comment]
This looks good to me @petermicuch .. I am waiting on @jackielii to confirm if the incorporated changes are good. 

Yes, I will hold off on #29 till this is merged. And I also need to work on a few updates to the Dockerfile in that PR.

**Comment by petermicuch** [comment]
@jackielii , @kmova is this now good to go? I have seen #29 was already merged to master.

**Comment by fabriciomrtnz** [comment]
> Thank you so much for the PR! It fixed our cluster for now. For anyone still using the (unfortunately deprecated but no replacement available) `stable/nfs-client-provisioner` helm chart, you can add the following to the helm values:
> 
> ```yaml
> image:
>   repository: rkevin/nfs-subdir-external-provisioner
>   tag: fix-k8s-1.20
> ```
> 
> This image is directly built on this PR. This is a temporary fix until this PR is merged upstream.

I do not use helm.

What does this translate to in the deployment-arm.yaml deployment?

**Comment by rkevin-arch** [comment]
> What does this translate to in the deployment-arm.yaml definition?

Try changing [line 24](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/blob/ba3360388e93114cf0c9237175a267b1cc67f874/deploy/deployment-arm.yaml#L24) from `image: quay.io/external_storage/nfs-client-provisioner-arm:latest` to `image: rkevin/nfs-subdir-external-provisioner:fix-k8s-1.20-arm` (or `image: rkevin/nfs-subdir-external-provisioner:fix-k8s-1.20` for x86_64). Note that this isn't an official image so you should be switching to an official one when it releases.

**Comment by groundhog2k** [comment]
I created a [fork of this repository](https://github.com/groundhog2k/nfs-subdir-external-provisioner) and used the patch from @petermicuch to create a [multi-arch image](https://hub.docker.com/r/groundhog2k/nfs-subdir-external-provisioner/tags?page=1&ordering=last_updated) of the nfs-subdir-external-provisioner. It works with Kubernetes 1.20.x on ARMv8 and AMD64 
Feel free to use it.

(I also copies the deprecated stable helm chart - to have a backup until a new offical chart will be available => https://github.com/groundhog2k/helm-charts/tree/master/charts/nfs-client-provisioner)

**Comment by ammmze** [comment]
Thank you @groundhog2k! I'm just getting my feet wet with kubernetes (just testing via helmfile and k3d ATM) to transition my home server from a docker-compose setup. And I appreciate the multi-arch image as well as I'm thinking I may pick up some raspberry pi's to start distributing my stuff across multiple hosts as well as keeping my AMD machine for GPU workloads.

**Comment by colfe34** [comment]
Guys thank you so much for this workaround, worked instantly.

**Comment by rkevin-arch** [comment]
I believe there is an official `nfs-subdir-external-provisioner` helm chart / image now, so you probably shouldn't be using the workarounds here. Better to switch to an official image.

**Pull request merged**